### DISCUSSION TITLE
Multiplayer implementation

### DISCUSSION THREAD

**Discussion by Praytic** [Praytic/youtd2#94] [Announcements]
## Discussion of the future multiplayer

How can we implement the multiplayer mode? 
1. Legacy one-player-one-base multiplayer. Players can just watch other players and can't affect each other's gameplays.
2. MazebertTD-like multiplayer. Up to 4 others can play on the same map, their tower is just shrunk to 1/4th size in the same square.

### Legacy multiplayer

**Pros:**
- Suitable for Race mode

**Cons:**
- Not a true multiplayer, players cannot interact with each other (only by chat)

### MazebertTD-like multiplayer

![](https://cdn.discordapp.com/attachments/1116928594850107483/1116954578898927616/image.png)

**Pros:**
- Allows cooperation between players in the same session
- Eliminates the need to make the map with 8 wave lanes

**Cons:**
- New balance calculations for such mode will be required.
- A single point of view for every game can be quite tedious for players, especially on this small map, especially for 240 waves-long games.
- LFG algorithm is more complicated than standard 2-player coop, because we should let the group of players decide how big is the party, what mode they want to play, pre-assembled groups, etc.
- Reduce towers in size, their effects and collisions might be not straightforward and may produce extra bugs.
- Screen with 4 players will have twice or trice more particles for one viewport. On bonus levels it can be a problem.
- Shared building space will produce race conditions, especially for 4 or more players.
- Not suitable for Race mode

**Comment by Praytic** [comment]
Good beginner tutorial video on how to implement multiplayer in Godot: https://www.youtube.com/watch?v=_ItA2r69c-Q

**Comment by Kvel2D** [comment]
Ended up implementing multiplayer like this:

- Co-op mode - 4 teams, 2 players in each team. Players in same team have different build spaces but can build towers near central lane so that they can clear waves together.
- FFA mode - 4 teams, 1 player per team. "Competitive" mode where each player gets one lane.
